ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -563,17 +567,109 @@ instance Serialise CC.Common.KeyHash where
+decodeByronBlock epochSlots = do
+  theBytes <- decodeUnknownCborDataItem
+  case decodeFullAnnotatedBytes "Block" internalDecoder (Lazy.fromStrict theBytes) of
+    Right it  -> pure $ ByronBlockOrEBB it
+    -- FIXME
+    --   err :: DecodeError
+    -- but AFAICT the only way to make the decoder fail is to give a `String`
+    -- to `fail`...
+    Left  err -> fail (show err)
+  where
+  internalDecoder :: Decoder s (CC.Block.ABlockOrBoundary ByteSpan)
+  internalDecoder = CC.Block.fromCBORABlockOrBoundary epochSlots
+
+-- | Get the encoded bytes of a header. A legacy Byron node (cardano-sl) would
+-- successfully decode a header from these.
+headerBytes :: Header (ByronBlockOrEBB cfg) -> Lazy.ByteString
+headerBytes blk = Lazy.pack [listLengthByte, discriminatorByte] `Lazy.append` Lazy.fromStrict mainBytes
+  where
+  listLengthByte = 0x82
+  discriminatorByte = case unByronHeaderOrEBB blk of
+    Left  _ -> 0x00
+    Right _ -> 0x01
+  mainBytes = case unByronHeaderOrEBB blk of
+    Left  ebb -> CC.Block.boundaryHeaderAnnotation ebb
+    Right hdr -> CC.Block.headerAnnotation hdr
+
+-- | Encode a header using CBOR-in-CBOR tag 24.
+encodeByronHeader :: Header (ByronBlockOrEBB cfg) -> Encoding
+encodeByronHeader = encodeUnknownCborDataItem . headerBytes
+
+-- | Inversion of `encodeByronHeader`. The annotation will be correct, because
+-- the full bytes are available thanks to the CBOR-in-CBOR encoding.
+decodeByronHeader :: CC.Slot.EpochSlots -> Decoder s (Header (ByronBlockOrEBB cfg))
+decodeByronHeader epochSlots = do
+  theBytes <- decodeUnknownCborDataItem
+  -- Would use decodeFullAnnotatedBytes, but we can't because it only works for
+  -- an f ByteSpan
+  case decodeFullAnnotatedBytes "Header" internalDecoder (Lazy.fromStrict theBytes) of
+    Right (LeftF ebb)  -> pure $ ByronHeaderOrEBB (Left ebb)
+    Right (RightF hdr) -> pure $ ByronHeaderOrEBB (Right hdr)
+    Left  err          -> fail (show err)
+  where
+  -- cardano-ledger does not export a decoder for genesis or main header with
+  -- the list length and tag bytes. It _does_ export one for blocks, used in
+  -- decodeByronBlock (fromCBORABlockOrBoundary)
+  internalDecoder :: Decoder s (EitherF CC.Block.ABoundaryHeader CC.Block.AHeader ByteSpan)
+  internalDecoder = fromEither <$> fromCBORAHeaderOrBoundary epochSlots
 
-encodeByronBlock :: Crypto.ProtocolMagicId -> CC.Slot.EpochSlots -> ByronBlockOrEBB cfg -> Encoding
-encodeByronBlock pm epochSlots bob = case unByronBlockOrEBB bob of
-  CC.Block.ABOBBlock b      -> CC.Block.toCBORABOBBlock epochSlots . void $ b
-  CC.Block.ABOBBoundary ebb -> CC.Block.toCBORABOBBoundary pm . void $ ebb
+-- | Defined only for use by decodeHeader.
+data EitherF g h t where
+  LeftF  :: g t -> EitherF g h t
+  RightF :: h t -> EitherF g h t
+
+instance (Functor g, Functor h) => Functor (EitherF g h) where
+  fmap f (LeftF g)  = LeftF  (fmap f g)
+  fmap f (RightF h) = RightF (fmap f h)
```

Comment:
I must be missing it somewhere. Where are we using the `EitherF`'s functor instance? It's the only reason to have this type.

Reply:
`decodeFullAnnotatedBytes` on line `650`.